### PR TITLE
feat  (logger): add loki compatibility.

### PR DESCRIPTION
**Pull request by IdrisDose**
Instead of using DataDog, due to pricing and other limitations I like to use Grafana Loki, I've added a compatibility patch for that within the logger import.

I had to do a hacky trick with the timestamp as Loki expects a nanosecond timestamp.

### CODE REVIEW

**Comment by thelindat** [comment]
I still want to do some additional logging changes in the future, but for now resolve the conflicts and I can merge.

File: imports/logger/server.lua [hunk 1]
```diff
@@ -27,4 +31,60 @@ if key ~= '' then
 	end
 end
 
+if lokiUser ~= '' and lokiKey ~= '' and lokiEndpoint ~= '' then
+    local site = ('https://%s:%s@%s/loki/api/v1/push'):format(lokiUser, lokiKey, lokiEndpoint)
+    local resourceName = GetCurrentResourceName()
+
+    function lib.logger(source, event, message, ...)
+        local timestamp = tostring(os.time(os.date("*t")) ) .. "000000000"
```
Comment (thelindat) [inline imports/logger/server.lua:39]:
```lua
('%s000000000'):format(os.time(os.date('*t')))
```
micro

**Review by thelindat — CHANGES_REQUESTED** [review]

File: imports/logger/server.lua [hunk 2]
```diff
@@ -27,4 +31,60 @@ if key ~= '' then
...
+            return
+        end
+		PerformHttpRequest(site, function(status, _, _, response)
+			if status ~= 204 then
+				-- Thanks, I hate it
+				print(('unable to submit logs to %s\n%s'):format(site, json.encode(response, {indent=true})))
+			end
+		end, 'POST', data, {
+			['Content-Type'] = 'application/json',
+		})
```
[the remark this answers is not in the document]
Comment (IdrisDose) [reply imports/logger/server.lua:85]: Yeah it's accepts arrays but in a weird way, will do though

File: imports/logger/server.lua [hunk 3]
```diff
@@ -27,4 +31,60 @@ if key ~= '' then
...
+                            service = event
+                        },
+                        values = {
+                            timestamp,
+                            json.encode(message)
+                        }
+                    }
+                }
+            })
+        end
```
[the remark this answers is not in the document]
Comment (IdrisDose) [reply imports/logger/server.lua:71]: Would you be okay if I made it an optional parameter? easier for log traversal and filtering if it's sent as a json kvp.

The intention behind this was to allow for both 'Player X given Player Y item Z' as well as `{player='X', target = 'Y', item='Z'}`

**Comment by IdrisDose** [comment]
Requires Testing before merging.

**Comment by thelindat** [comment]
> Requires Testing before merging.

Hope you mean you'll be testing it, because I'd like to merge this before a new tag but need to setup grafana and whatnot.
Requires documentation changes as well https://github.com/overextended/overextended.github.io/blob/main/docs/ox_lib/Logger/Server.md

**Comment by IdrisDose** [comment]
Yeah, I'll be testing it after work today.

**Comment by IdrisDose** [comment]
Tested and works

**Comment by thelindat** [comment]
Either two separate pages or use tabs, don't mind. I need to do a full guide on datadog anyway.

**Pull request merged**